### DISCUSSION TITLE
Symfony's current version has a bug and is not possible to get the $container variable

### DISCUSSION THREAD

**Discussion by ghost** [symfony/symfony#42837] [Q&A]
The $container variable is mentioned dozens of times throughout the documentation.

However, when trying to access it from a controller in the latest versions, it gives the following error message:

"Notice: Undefined variable: container"

The message happens because of course the variable is undefined. However, there is no way of getting the service container in the latest Symfony releases.

Instead of acknowledging the issue, the maintainer "derrabus", who is answering for multiple broken and undocumented repositories throughout the Symfony ecosystem, is closing every issue that is opened by myself and others related to this problem and trying to sweep it under the rug.

Every Symfony library that uses the service container system is broken because the $container variable is inaccessible. Multiple repositories are affected and multiple Symfony functionalities are broken because of this.

The $container variable, mentioned throughout the documentation of multiple repositories, is not accessible anymore in the latest Symfony releases.

**Comment by ahundiak** [comment]
Please provide a specific example  of where the documents indicate that $container should be accessible.  An actual link if possible.  

And keep in mind that the usage of the container has evolved over time.  Nowadays, most services are not accessible via the container even if you have the container itself.  When reading the documentation, it always important to keep in mind which specific version you are using.

**Comment by ghost** [comment]
THANK YOU!

The $container variable is mentioned multiple and multiple times in the docs, without ever there having a description of how to access them. For example, in the page about Service Containers: https://symfony.com/doc/current/service_container.html. A search for "$container" shows all apparitions.

Despite most services not being accessible via the container, there are still libraries and bundles that mention the $container variable as the only possible way to use those libraries. For example, one of the main serializers compatible with Doctrine Entities:

https://jmsyst.com/bundles/JMSSerializerBundle

Now this could be considered a pretty fundamental piece of software for the Symfony ecosystem. Other libraries also mention $container without ever telling how to get it.

Now, I guessed and guessed and I could find $this->container from the controllers. However, as you mentioned most services are not accessible from there, and an error message appears telling that only the some of the most important service containers are kept.

Now the question remains about how to access services and bundles libraries (which are documented as being accessible through the container). I tried parameter hinting with the corresponding classes (for example, JMS\SerializerBundle\JMSSerializerBundle) and it also didn't work.

I cannot 100% say this whole situation is a bug, if you use a very pedantic definition of a bug. It might be that there is a way in a vast space of possible guesses to make things work, it is just not possible for any developer to know how to make it work in a lifetime, since the documented interfaces don't work as documented. Or perhaps they might read or even modify the Symfony code and make things work. That would be what I and most people call a bug... If there are interfaces that are documented in a given way and they don't work as documented anymore, this would be a bug, not a documentation issue...

I personally tried a lot of workaround, but each of them leads to another issue or another thing that doesn't work as documented and it becomes an infinite cycle.

**Comment by ahundiak** [comment]
Symfony documentation has a well deserved reputation for being confusing.  I can sympathize.  Being able to read the source code is pretty much a necessity.  Looking at the AbstractController class will, for example, reveal the existence of a protected $container variable.  Which of course is not the global container and is only a service locator capable of accessing only a small subset of all services.  Writing and maintaing useful docs is far more difficult than writing code and not nearly as much fun.

It does take some time for dependency injection to 'sink in'.  In the early days you did have access to the container wherever you needed it it and you were encouraged to pull services directly from it.  Nowadays the process is to inject specific services as needed.  Which may sound simple enough but there are various gotchas especially when using autowire.

Hence the request for a specific example.  If you can present a straight forward 'how do I get access to this service from this class' question then you may get some useful help.  Stackoverflow has a gazillion questions of that nature.

And I don't think the JMSSerializerBundle is considered to be part of the core ecosystem anymore.  The Serializer component itself is now powerful enough to handle most use cases.

**Comment by ghost** [comment]
"The Serializer component itself is now powerful enough to handle most use cases."

Fortunately I could get things working with it. I wrote a lot of custom code and my own BaseController class inheriting from AbstractController and now I can access most of the stuff I need and have control about what is going on.

The JMSSerializerBundle would be an example, though, of a service that was usually got through the $container variable, and now seems impossible to access. Whether it is a bug or not, I think is something that should be investigated further before classifying it as just a documentation issue or a support question. Was the intention of Symfony developers to deprecate libraries that worked through the container variable? It doesn't seem like it, or at least the intention was not publicly stated (since not even the Service Container doc removed the mentions to $container). If not, do these libraries still work somehow? How? The JMSSerializerBundle doesn't seem to be working through dependency injection, and at this point I really don't know whether it is a JMSSerializerBundle or Symfony bug. Both are at least even in that neither have updated their docs since the $container variable stopped working to get services.

**Comment by ahundiak** [comment]
Been some time since I used the JMS bundle.  I made a new symfony 5.3 app and loaded the bundle:
```
symfony new --full jms; cd jms
composer require jms/serializer-bundle

bin/console debug:container JMS
...lots of services...
[54] JMS\Serializer\SerializerInterface
...
54
Service ID       jms_serializer.serializer  
Class            JMS\Serializer\Serializer  

Type hinted against the alias
    public function __construct(private \JMS\Serializer\SerializerInterface $jms)
    {
        parent::__construct();
    }
```
And everything seemed to work as expected.  The trick for using autoload it to find an Interface to typehint against.  debug:container can be very helpful.  If you can't find something to typehint against then you can use bind or something similar.

The bundle itself seems to be quite dated though there does seem to be activity on github.

If you use this service quite a bit in controllers then just override AbstractController::getSubscribedServices and add it.

You might also be amused to discover that in 6.0, even more services will be removed from the AbstractController.  The doctrine service, for example, will now need to be injected.  I don't know if you are aware of this but the ['Living on the Edge'](https://symfony.com/blog/category/living-on-the-edge) blog discusses feature changes.  It is quite useful for keeping track of what is going on in Symfony.
